pkg/master/controller.go

```diff
@@ -271,38 +286,80 @@ func (c *Controller) SetEndpoints(serviceName string, ip net.IP, endpointPorts [
 				}
 			}
 		}
-		return c.EndpointRegistry.UpdateEndpoints(ctx, e)
 	}
-	// We didn't make any changes, no need to actually call update.
-	return nil
+	if !portsCorrect {
+		// Reset ports.
+		e.Subsets[0].Ports = endpointPorts
+	}
+	glog.Warningf("Resetting endpoints for master service %q to %v", serviceName, e)
+	return c.EndpointRegistry.UpdateEndpoints(ctx, e)
 }
 
-// Determine if the endpoint is in the format SetEndpoints expect (one subset,
+// Determine if the endpoint is in the format ReconcileEndpoints expect (one subset,
 // correct ports, N IP addresses); and if the specified IP address is present and
 // the correct number of ip addresses are found.
-func checkEndpointSubsetFormat(e *api.Endpoints, ip string, ports []api.EndpointPort, count int) (formatCorrect, ipCorrect bool) {
+func checkEndpointSubsetFormat(e *api.Endpoints, ip string, ports []api.EndpointPort, count int, reconcilePorts bool) (formatCorrect bool, ipCorrect bool, portsCorrect bool) {
 	if len(e.Subsets) != 1 {
-		return false, false
+		return false, true, true
 	}
 	sub := &e.Subsets[0]
-	if len(sub.Ports) != len(ports) {
-		return false, false
-	}
-	for _, port := range ports {
-		contains := false
-		for _, subPort := range sub.Ports {
-			if port == subPort {
-				contains = true
-			}
+	if reconcilePorts {
+		if len(sub.Ports) != len(ports) {
+			return true, false, false
 		}
-		if !contains {
-			return false, false
+		for i, port := range ports {
+			if port != sub.Ports[i] {
+				return true, false, false
```

Comment:
add a test case to make sure incorrect ports, but a correct IP, doesn't result in a double-appended IP